src/js/mqtt.js

```diff
@@ -0,0 +1,308 @@
+   * Qos level 2
+   * Handle PUBCOMP package. If this package is arrived, the sending process
+   * is done.
+   */
+  this.emit('pubcomp', data);
+};
+
+MQTTClient.prototype.onsuback = function(data) {
+  /*
+   * Successful subscription, the client will get messages from the requested
+   * topic. The granted QoS is given in data.
+   */
+};
+
+MQTTClient.prototype.onerror = function(error) {
+  this.emit('error', error);
+};
+
+MQTTClient.prototype.onmessage = function(message, topic, qos, packet_id) {
+  var data = {
+    message: message,
+    topic: topic,
+    qos: qos,
+    packet_id: packet_id,
+  };
+
+  if (qos == 1) {
+    var options = {
+      type: PacketTypeEnum.PUBACK,
+      packet_id: packet_id
+    }
+
+    this.sendAcknowledge(options);
+  } else if (qos == 2) {
+    var options = {
+      type: PacketTypeEnum.PUBREC,
+      packet_id: packet_id
+    }
+    this.sendAcknowledge(options);
+  }
+
+  this.emit('message', data);
+};
+
+MQTTClient.prototype.publish = function(options) {
+  if (!Buffer.isBuffer(options.message)) {
+    options.message = new Buffer(options.message);
+  }
+  if (!Buffer.isBuffer(options.topic)) {
+    options.topic = new Buffer(options.topic);
+  }
+
+  if (util.isNumber(options.qos) && options.qos > 0) {
+    options.packet_id = this._package_id;
+    this._package_id++;
+
+    var buffer = native.publish(options);
+    this._socket.write(buffer);
+
+    var self = this;
+
+    var interval = setInterval(function() {
+      self._socket.write(buffer);
+    }, 3000);
+
+    this.on('puback', function(data) {
+      clearInterval(interval);
+    });
+    this.on('pubrec', function(data) {
+      clearInterval(interval);
+    });
+
+    return;
+  }
+
+  this._socket.write(native.publish(options));
+};
+
+MQTTClient.prototype.subscribe = function(options) {
+  if (!Buffer.isBuffer(options.topic)) {
+    options.topic = new Buffer(options.topic);
+  }
+
+  var buff = native.subscribe(options);
+  this._socket.write(buff);
+};
+
+MQTTClient.prototype.ping = function() {
+  var buff = native.ping();
+  this._socket.write(buff);
+};
+
+MQTTClient.prototype.unsubscribe = function(topic) {
+  if (!Buffer.isBuffer(topic)) {
+    topic = new Buffer(topic);
+  }
+
+  var buf = native.unsubscribe(topic);
+  this._socket.write(buf);
+}
+
+MQTTClient.prototype.onpingresp = function() {
+  this.emit('onpingresp');
+}
+
+MQTTClient.prototype.sendAcknowledge = function(options) {
+  var buff = native.sendAck(options);
+  this._socket.write(buff);
+}
+
+/*
+ * Returns an unique client ID based on current time.
+ */
+function defaultClientId() {
+  return 'iotjs_mqtt_client_' + Date.now();
+}
+
+function getClient(connectionOptions) {
+  if (util.isUndefined(connectionOptions.clientId)) {
+    connectionOptions.clientId = defaultClientId();
+  }
+  if (!Buffer.isBuffer(connectionOptions.clientId)) {
+    connectionOptions.clientId = new Buffer(connectionOptions.clientId);
+  }
+  return MQTTClient(connectionOptions);
```

Comment:
`return new MQTTClient(connectionOptions);`

Reply:
Yeah about the prototype thing, we haven't refactored the JS code yet, were focusing on the C side. 
We were planning on changing them as you mentioned it.